File: starter-packs/zig/src/hlt.zig
pub const DOCK_RADIUS: f64 = 4.0;
pub const SHIP_RADIUS: f64 = 0.5;
pub const MAX_SPEED: u8 = 7;

pub const DockStatus = enum(u8) { undocked = 0, docking = 1, docked = 2, undocking = 3 };

pub const Ship = struct {
    id: u32,
    owner_id: u32,
    x: f64,
    y: f64,
    health: u8,
    status: DockStatus,
    planet_id: u32,
    docking_progress: u8,

    pub fn is_undocked(self: Ship) bool {
        return self.status == .undocked;
    }

    pub fn distance_to_planet(self: Ship, p: Planet) f64 {
        return @sqrt((p.x - self.x) * (p.x - self.x) + (p.y - self.y) * (p.y - self.y));
    }

    pub fn angle_to_planet(self: Ship, p: Planet) u16 {
        const deg = std.math.atan2(p.y - self.y, p.x - self.x) * (180.0 / std.math.pi);
        return @intFromFloat(@mod(deg + 360.0, 360.0));
    }

    pub fn can_dock(self: Ship, p: Planet) bool {
        return self.distance_to_planet(p) <= p.size + DOCK_RADIUS + SHIP_RADIUS;
    }
};

pub const Planet = struct {
    id: u32,
    x: f64,
    y: f64,
    size: f64,
    halite: f64,
    owner_id: u32 = 0,
    docked_count: u32 = 0,

    pub fn docking_spots(self: Planet) u32 {
        return @intFromFloat(@floor(self.size));
    }

    pub fn is_full(self: Planet) bool {
        return self.docked_count >= self.docking_spots();
    }
};

pub const Player = struct {
    id: u32,
    ships: std.AutoHashMap(u32, Ship),
};

pub const GameMap = struct {
    width: u32,
    height: u32,
    my_id: u32,
    players: []Player,
    planets: std.AutoHashMap(u32, Planet),
    allocator: std.mem.Allocator,

    pub fn me(self: *GameMap) ?*Player {
        for (self.players) |*p| {
            if (p.id == self.my_id) return p;
        }
        return null;
    }

    pub fn deinit(self: *GameMap) void {
        for (self.players) |*p| p.ships.deinit();
        self.allocator.free(self.players);
        self.planets.deinit();
    }
};

pub const Game = struct {
    io: std.Io,
    stdin: std.Io.File,
    stdout: std.Io.File,
    my_id: u32,
    map: GameMap,
    // 32 KB — enough for a state line with ~200 ships and 5 planets
    in_buf: [32768]u8 = undefined,
    out_buf: [512]u8 = undefined,

    pub fn init(io: std.Io, name: []const u8, allocator: std.mem.Allocator) !Game {
        var stdin = std.Io.File.stdin();
        var stdout = std.Io.File.stdout();
        var in_buf: [32768]u8 = undefined;
        var out_buf: [512]u8 = undefined;
        var r = stdin.readerStreaming(io, &in_buf);
        var w = stdout.writerStreaming(io, &out_buf);

        // Line 1: my_player_id
        const l1 = try r.interface.takeDelimiterExclusive('\n');
        r.interface.toss(1);
        const my_id = try std.fmt.parseInt(u32, std.mem.trimEnd(u8, l1, "\r"), 10);

        // Line 2: width height
        const l2 = try r.interface.takeDelimiterExclusive('\n');
        r.interface.toss(1);
        var it2 = std.mem.tokenizeScalar(u8, l2, ' ');
        const width = try std.fmt.parseInt(u32, it2.next().?, 10);
        const height = try std.fmt.parseInt(u32, it2.next().?, 10);

        // Line 3: full game state
        const l3 = try r.interface.takeDelimiterExclusive('\n');
        r.interface.toss(1);
        const state = try parseState(l3, allocator);

        // Send bot name
        try w.interface.print("{s}\n", .{name});
        try w.flush();

        return .{
            .io = io,
            .stdin = stdin,
            .stdout = stdout,
            .my_id = my_id,
            .map = .{
                .width = width,
                .height = height,
                .my_id = my_id,
                .players = state.players,
                .planets = state.planets,
                .allocator = allocator,
            },
            .in_buf = in_buf,
            .out_buf = out_buf,
        };
    }

    pub fn update_map(self: *Game) !*GameMap {
        var r = self.stdin.readerStreaming(self.io, &self.in_buf);
        const line = try r.interface.takeDelimiterExclusive('\n');
        r.interface.toss(1);

        // Replace map contents with fresh parse
        for (self.map.players) |*p| p.ships.deinit();
        self.map.allocator.free(self.map.players);
        self.map.planets.deinit();

        const state = try parseState(line, self.map.allocator);
        self.map.players = state.players;
        self.map.planets = state.planets;

        return &self.map;
    }

    pub fn send_commands(self: *Game, cmds: []const []const u8) !void {
        var w = self.stdout.writerStreaming(self.io, &self.out_buf);
        for (cmds, 0..) |cmd, i| {
            if (i > 0) try w.interface.writeAll(" ");
            try w.interface.writeAll(cmd);
        }
        try w.interface.writeAll("\n");
        try w.flush();
    }
};

const ParseState = struct {
    players: []Player,
    planets: std.AutoHashMap(u32, Planet),
};

fn parseState(line: []const u8, allocator: std.mem.Allocator) !ParseState {
    var it = std.mem.tokenizeAny(u8, line, " \r");

    const n_players = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
    const players = try allocator.alloc(Player, n_players);
    errdefer allocator.free(players);

    for (players) |*player| {
        const pid = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        const n_ships = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        player.* = .{ .id = pid, .ships = std.AutoHashMap(u32, Ship).init(allocator) };

        for (0..n_ships) |_| {
            const sid = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
            const x = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
            const y = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
            const hp = try std.fmt.parseInt(u8, it.next() orelse return error.BadState, 10);
            _ = it.next(); // vel_x
            _ = it.next(); // vel_y
            const docked = try std.fmt.parseInt(u8, it.next() orelse return error.BadState, 10);
            const planet_id = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
            const progress = try std.fmt.parseInt(u8, it.next() orelse return error.BadState, 10);
            _ = it.next(); // weapon cooldown
            try player.ships.put(sid, .{
                .id = sid,
                .owner_id = pid,
                .x = x,
                .y = y,
                .health = hp,
                .status = @enumFromInt(docked),
                .planet_id = planet_id,
                .docking_progress = progress,
            });
        }
    }

    const n_planets = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
    var planets = std.AutoHashMap(u32, Planet).init(allocator);
    errdefer planets.deinit();

    for (0..n_planets) |_| {
        const pid = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        const x = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
        const y = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
        _ = it.next(); // planet hp (255)
        const size = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
        _ = it.next(); // docking spots
        _ = it.next(); // current production
        const halite = try std.fmt.parseFloat(f64, it.next() orelse return error.BadState);
        const owned = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        const owner_id = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        const n_docked = try std.fmt.parseInt(u32, it.next() orelse return error.BadState, 10);
        for (0..n_docked) |_| _ = it.next(); // docked ship ids
        try planets.put(pid, .{
            .id = pid,
            .x = x,
            .y = y,
            .size = size,
            .halite = halite,
            .owner_id = if (owned != 0) owner_id else 0,
            .docked_count = n_docked,
        });
    }

    return .{ .players = players, .planets = planets };
}

pub fn cmd_thrust(buf: []u8, ship_id: u32, mag: u8, angle: u16) []u8 {
    return std.fmt.bufPrint(buf, "t {d} {d} {d}", .{ ship_id, @min(MAX_SPEED, mag), angle % 360 }) catch unreachable;
}

pub fn cmd_dock(buf: []u8, ship_id: u32, planet_id: u32) []u8 {
    return std.fmt.bufPrint(buf, "d {d} {d}", .{ ship_id, planet_id }) catch unreachable;
}

pub fn cmd_undock(buf: []u8, ship_id: u32) []u8 {
    return std.fmt.bufPrint(buf, "u {d}", .{ship_id}) catch unreachable;
}

const std = @import("std");

File: starter-packs/rust/src/hlt.rs
use std::collections::HashMap;
use std::io::{self, BufRead, Write};

pub const DOCK_RADIUS: f64 = 4.0;
pub const SHIP_RADIUS: f64 = 0.5;
pub const MAX_SPEED: u32   = 7;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DockStatus { Undocked = 0, Docking = 1, Docked = 2, Undocking = 3 }

impl From<u32> for DockStatus {
    fn from(v: u32) -> Self {
        match v { 1 => Self::Docking, 2 => Self::Docked, 3 => Self::Undocking, _ => Self::Undocked }
    }
}

#[derive(Debug, Clone)]
pub struct Ship {
    pub id: u32, pub owner_id: u32,
    pub x: f64, pub y: f64, pub health: u32,
    pub status: DockStatus, pub planet_id: u32, pub docking_progress: u32,
}

impl Ship {
    pub fn is_undocked(&self) -> bool { self.status == DockStatus::Undocked }

    pub fn distance_to(&self, p: &Planet) -> f64 {
        ((p.x - self.x).powi(2) + (p.y - self.y).powi(2)).sqrt()
    }

    pub fn angle_to(&self, p: &Planet) -> u32 {
        let deg = (p.y - self.y).atan2(p.x - self.x).to_degrees();
        ((deg % 360.0 + 360.0) as u32) % 360
    }

    pub fn can_dock(&self, p: &Planet) -> bool {
        self.distance_to(p) <= p.size + DOCK_RADIUS + SHIP_RADIUS
    }
}

#[derive(Debug, Clone)]
pub struct Planet {
    pub id: u32,
    pub x: f64, pub y: f64, pub size: f64, pub halite: f64,
    pub owner_id: u32, pub docked_count: u32,
}

impl Planet {
    pub fn docking_spots(&self) -> u32 { self.size.floor() as u32 }
    pub fn is_full(&self) -> bool { self.docked_count >= self.docking_spots() }
}

#[derive(Debug)]
pub struct Player {
    pub id: u32,
    pub ships: HashMap<u32, Ship>,
}

pub struct GameMap {
    pub width: u32, pub height: u32,
    pub players: Vec<Player>,
    pub planets: HashMap<u32, Planet>,
    pub my_id: u32,
}

impl GameMap {
    pub fn me(&self) -> Option<&Player> {
        self.players.iter().find(|p| p.id == self.my_id)
    }
}

pub struct Game {
    pub map: GameMap,
    reader: io::BufReader<io::Stdin>,
    buf: String,
}

impl Game {
    pub fn new(name: &str) -> Self {
        let mut reader = io::BufReader::new(io::stdin());
        let mut buf = String::new();

        let mut next_line = |r: &mut io::BufReader<io::Stdin>, b: &mut String| -> String {
            b.clear();
            r.read_line(b).unwrap();
            b.trim_end().to_owned()
        };

        let l1 = next_line(&mut reader, &mut buf);
        let my_id: u32 = l1.trim().parse().unwrap();

        let l2 = next_line(&mut reader, &mut buf);
        let mut it2 = l2.split_whitespace();
        let width:  u32 = it2.next().unwrap().parse().unwrap();
        let height: u32 = it2.next().unwrap().parse().unwrap();

        let l3 = next_line(&mut reader, &mut buf);
        let (players, planets) = parse_state(&l3);

        print!("{}\n", name);
        io::stdout().flush().unwrap();

        Game {
            map: GameMap { width, height, players, planets, my_id },
            reader,
            buf,
        }
    }

    pub fn update_map(&mut self) -> &GameMap {
        self.buf.clear();
        self.reader.read_line(&mut self.buf).unwrap();
        let line = self.buf.trim_end().to_owned();
        let (players, planets) = parse_state(&line);
        self.map.players = players;
        self.map.planets = planets;
        &self.map
    }

    pub fn send_commands(&self, cmds: &[String]) {
        println!("{}", cmds.join(" "));
        io::stdout().flush().unwrap();
    }
}

fn parse_state(line: &str) -> (Vec<Player>, HashMap<u32, Planet>) {
    let tokens: Vec<&str> = line.split_whitespace().collect();
    let mut pos = 0;

    macro_rules! take_u32 {
        () => {{ let v: u32 = tokens[pos].parse().unwrap(); pos += 1; v }};
    }
    macro_rules! take_f64 {
        () => {{ let v: f64 = tokens[pos].parse().unwrap(); pos += 1; v }};
    }
    macro_rules! skip {
        () => {{ pos += 1; }};
    }

    let n_players = take_u32!() as usize;
    let mut players = Vec::with_capacity(n_players);
    for _ in 0..n_players {
        let pid     = take_u32!();
        let n_ships = take_u32!() as usize;
        let mut ships = HashMap::new();
        for _ in 0..n_ships {
            let sid      = take_u32!();
            let x        = take_f64!();
            let y        = take_f64!();
            let hp       = take_u32!();
            skip!(); skip!();           // vel_x, vel_y
            let docked   = take_u32!();
            let planet_id = take_u32!();
            let progress  = take_u32!();
            skip!();                    // weapon cooldown
            ships.insert(sid, Ship {
                id: sid, owner_id: pid,
                x, y, health: hp,
                status: docked.into(), planet_id, docking_progress: progress,
            });
        }
        players.push(Player { id: pid, ships });
    }

    let n_planets = take_u32!() as usize;
    let mut planets = HashMap::new();
    for _ in 0..n_planets {
        let pid      = take_u32!();
        let x        = take_f64!();
        let y        = take_f64!();
        skip!();                        // planet hp (255)
        let size     = take_f64!();
        skip!();                        // docking spots
        skip!();                        // current production
        let halite   = take_f64!();
        let owned    = take_u32!();
        let owner_id = take_u32!();
        let n_docked = take_u32!();
        for _ in 0..n_docked {
            skip!();                    // docked ship ids
        }
        planets.insert(pid, Planet {
            id: pid, x, y, size, halite,
            owner_id: if owned != 0 { owner_id } else { 0 },
            docked_count: n_docked,
        });
    }

    (players, planets)
}

pub fn thrust(ship_id: u32, magnitude: u32, angle: u32) -> String {
    format!("t {} {} {}", ship_id, magnitude.min(MAX_SPEED), angle % 360)
}

pub fn dock(ship_id: u32, planet_id: u32) -> String {
    format!("d {} {}", ship_id, planet_id)
}

pub fn undock(ship_id: u32) -> String {
    format!("u {}", ship_id)
}
